yarn-project/acir-simulator/src/avm/avm_context.ts

```diff
@@ -1,140 +1,63 @@
+   * Create a new AVM context
+   * @param worldState - Manages mutable state during execution - (caching, fetching)
+   * @param environment - Contains constant variables provided by the kernel
+   * @param machineState - VM state that is modified on an instruction-by-instruction basis
+   * @returns new AvmContext instance
    */
-  public static newWithForkedState(executionEnvironment: AvmExecutionEnvironment, journal: AvmJournal): AvmContext {
-    const forkedState = AvmJournal.branchParent(journal);
-    return new AvmContext(executionEnvironment, forkedState);
-  }
+  constructor(
+    public worldState: AvmWorldStateJournal,
+    public environment: AvmExecutionEnvironment,
+    public machineState: AvmMachineState,
+  ) {}
 
   /**
-   * Prepare a new AVM context that will be ready for an external call
-   * - It will fork the journal
-   * - It will set the correct execution Environment Variables for a call
-   *    - Alter both address and storageAddress
+   * Prepare a new AVM context that will be ready for an external/nested call
+   * - Fork the world state journal
+   * - Derive a machine state from the current state
+   *   - E.g., gas metering is preserved but pc is reset
+   * - Derive an execution environment from the caller/parent
+   *   - Alter both address and storageAddress
    *
-   * @param address - The contract to call
-   * @param executionEnvironment - The current execution environment
-   * @param journal - The current journal
+   * @param address - The contract instance to initialize a context for
+   * @param calldata - Data/arguments for nested call
    * @returns new AvmContext instance
    */
-  public static prepExternalCallContext(
-    address: AztecAddress,
-    calldata: Fr[],
-    executionEnvironment: AvmExecutionEnvironment,
-    journal: AvmJournal,
-  ): AvmContext {
-    const newExecutionEnvironment = executionEnvironment.newCall(address, calldata);
-    const forkedState = AvmJournal.branchParent(journal);
-    return new AvmContext(newExecutionEnvironment, forkedState);
+  public createNestedContractCallContext(address: AztecAddress, calldata: Fr[]): AvmContext {
+    const newExecutionEnvironment = this.environment.deriveEnvironmentForNestedCall(address, calldata);
+    const forkedWorldState = this.worldState.fork();
+    const machineState = AvmMachineState.fromState(this.machineState);
+    return new AvmContext(forkedWorldState, newExecutionEnvironment, machineState);
   }
 
   /**
-   * Prepare a new AVM context that will be ready for an external static call
-   * - It will fork the journal
-   * - It will set the correct execution Environment Variables for a call
-   *    - Alter both address and storageAddress
+   * Prepare a new AVM context that will be ready for an external/nested static call
+   * - Fork the world state journal
+   * - Derive a machine state from the current state
+   *   - E.g., gas metering is preserved but pc is reset
```

Comment:
Gas metering actually is not preserved. You allocate gas to a nested context via the call instruction's gas argument.